### PR TITLE
Use replay before deploying a new version of a Workflow

### PR DESCRIPTION
**Pull request by rachfop**
## What does this PR do?

- EDU-181

### CODE REVIEW

File: kb/deploy-new-version-of-workflow.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+---
+slug: deploy-new-version-of-workflow
+title: Use Replay before deploy a new version of a Workflow
+tags:
+  - troubleshooting
+  - how-to
+date: 2023-05-11T00:00:00Z
+---
+
+When making code changes to a Workflow, it is important to avoid introducing non-deterministic behavior; otherwise, you may encounter `NonDeterministicWorkflowError` exceptions in your Workflow History.
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:10]: Not sure this error is what it's called in every SDK

File: kb/deploy-new-version-of-workflow.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+One way to prevent this exception from occurring is to use the [Workflow Replayer](/workflows#replays) to replay existing Workflow Histories against your new code before deploying it to production.
+
+If the Replay is successful, you can be confident that your new Workflow version is compatible with your existing Workflow History.
+
+## How to use the Workflow Replayer
+
+To use the Workflow Replayer, you need to provide the following:
+
+- The Workflow Replayer to register the Workflow that is going to be replayed
+- A Workflow History to execute a single Workflow Task from the JSON History file
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:21]:
```suggestion
- A Workflow History to execute (can be from the JSON History file or loaded directly from the client)
```

File: kb/deploy-new-version-of-workflow.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+- (optional) Replay Workflow from a specific point in the History
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:22]: Most SDKs don't support this, we should not include it in the KB

File: kb/deploy-new-version-of-workflow.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+You can load the Event History from a JSON file on disk, obtained through one of the following methods:
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:24]: Don't have to load from JSON file on disk, can just use client

File: kb/deploy-new-version-of-workflow.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+### Programmatically
+
+The Workflow Replayer is available in each SDK.
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:28]: A workflow replayer being available in each SDK seems unrelated to this section which is about loading event history programmatically

File: kb/deploy-new-version-of-workflow.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+1. Navigate to the Workflow Execution you want to download the History from.
+2. In the **Recent Events** section, select **Download**.
+
+Once you have the Workflow JSON History file, you can use the Workflow Replayer to replay the Workflow History.
+
+### Replay existing Workflow Histories in CI
+
+Use your existing CI pipeline to replay Workflow Histories against your new Workflow versions to ensure that they are compatible with every change pushed to your version control management.
+You can replay existing Workflow Histories or partial Workflow Histories in your CI pipeline and validate that they do not break determinism and prevent failures from occurring in production.
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:51]:
```suggestion
You can replay existing Workflow Histories in your CI pipeline and validate that they do not break determinism and prevent failures from occurring in production.
```

Should not discuss partial histories, approach not implemented in every SDK

File: kb/deploy-new-version-of-workflow.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+### The Temporal Web UI
+
+Navigate through the Temporal Web UI to download the Workflow History file:
+
+1. Navigate to the Workflow Execution you want to download the History from.
+2. In the **Recent Events** section, select **Download**.
+
+Once you have the Workflow JSON History file, you can use the Workflow Replayer to replay the Workflow History.
+
+### Replay existing Workflow Histories in CI
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:48]: Are these `###` level headings about obtaining event history? Should this be at a different heading level or under a different section?

File: kb/deploy-new-version-of-workflow.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+Use your existing CI pipeline to replay Workflow Histories against your new Workflow versions to ensure that they are compatible with every change pushed to your version control management.
+You can replay existing Workflow Histories or partial Workflow Histories in your CI pipeline and validate that they do not break determinism and prevent failures from occurring in production.
+
+For more information on testing Workflows, see [Test frameworks](/application-development/testing#test-frameworks)
+
+## What happens if I don't use the Workflow Replayer?
+
+If you don't use the Workflow Replayer, you may see `NonDeterministicWorkflowError` exceptions in your Workflow History. This means that your new code is not compatible with existing Workflow Histories.
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:57]: You may see this even if you do use the replayer and it succeeds. The nature of non-determinism is you don't know _when_ it will decide to be non-deterministic. We should not act as if using a replayer is a panacea for determinism prevention, it is but one tool that _may_ help, but may not.

File: kb/deploy-new-version-of-workflow.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+To fix this, you might update your Workflow code to be backwards compatible with the existing Workflow Histories, then reset the Workflow, which can be painful for many stuck executions.
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:58]: Not sure a sentence about how to recover from non-deterministic errors is best here. Or if it is, make it clear that's what is being discussed (at least I found it tough at first to understand you were talking about fixing stuck workflows errors with "this").

Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:58]: You may not need to reset your workflow.

File: kb/deploy-new-version-of-workflow.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+To use the Workflow Replayer, you need to provide the following:
+
+- The Workflow Replayer to register the Workflow that is going to be replayed
+- A Workflow History to execute, either:
+  - Loaded directly from the Client
+  - Loaded from the JSON History file
+
+If you choose to load the Workflow History directly from the Client, consult the [SDK specific documentation](/application-development/testing#replay) for more information.
+
+If you chose to load the Event History from a JSON file on disk, use one the following methods:
```
Comment (cretz) [inline kb/deploy-new-version-of-workflow.md:31]: This says from JSON file on disk and then the first subsection is unrelated to JSON, files, or disk

File: kb/deploy-new-version-of-workflow.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+1. Navigate to the Workflow Execution you want to download the History from.
+2. In the **Recent Events** section, select **Download**.
+
+Once you have the Workflow JSON History file, you can use the Workflow Replayer to replay the Workflow History, in the SDK of your choice.
+
+## Replay existing Workflow Histories in CI
+
+Use your existing CI pipeline to replay Workflow Histories against your new Workflow versions to ensure that they are compatible with every change pushed to your version control management.
+You can replay existing Workflow Histories in your CI pipeline and validate that they do not break determinism and prevent failures from occurring in production.
```
Comment (Sushisource) [inline kb/deploy-new-version-of-workflow.md:57]: Would be good here to mention that a good technique can be to fetch the most recent `N` histories from the task queue you care about testing (some open, some closed) and replay all of them to gain confidence

File: kb/deploy-new-version-of-workflow.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+### The Temporal Web UI
+
+Navigate through the Temporal Web UI to download the Workflow History file:
+
+1. Navigate to the Workflow Execution you want to download the History from.
+2. In the **Recent Events** section, select **Download**.
+
+Once you have the Workflow JSON History file, you can use the Workflow Replayer to replay the Workflow History, in the SDK of your choice.
+
+## Replay existing Workflow Histories in CI
```
Comment (flossypurse) [inline kb/deploy-new-version-of-workflow.md:50]: This is core documentation that should live here: https://docs.temporal.io/workflows#replays

This explains a significant use case for the Replayer which is a key aspect of explaining it as a concept -

**Review by flossypurse — CHANGES_REQUESTED** [review]
This feels like a bandaid for a topic that needs surgery.

I think I want to see this more expertly integrated into the docs -

**Review by cretz — DISMISSED** [review]
